### PR TITLE
VUU-27 remotePersistenceManager implementation

### PR DESCRIPTION
**Pull request by vferraro-scottlogic**
- [x] verify against server

### Changes
- RemoteLayoutPersistenceManager implementation 
- rename "date" to created in LayoutMetadata type
- change return type for createLayout in LayoutPersistenceManager to Promise<Metadata>
- edit LocalLayoutPersistenceManager.createLayout to return Metadata
- add unit tests for RemoteLayoutPersistenceManager 
- change unit tests for LocalLayoutPersistenceManager
- add catching layer to LayoutManagerProvider()
- introduce environment variable for switching between local and remote layout manager

### CODE REVIEW

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  loadLayout(id: string): Promise<LayoutJSON> {
+    return new Promise((resolve, reject) => {
+      fetch(`${baseURL}/${layoutsSaveLocation}/${id}`, {})
+        .then((response) => {
+          if (!response.ok) {
+            reject(new Error(response.statusText));
+          }
+          response.json().then((layout) => {
+            if (!layout) {
+              reject(new Error("invalid layout"));
```
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:96]:
```suggestion
reject(new Error("Response did not contain a valid layout"));
```

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  createLayout(
+    metadata: Omit<LayoutMetadata, "id" | "created">,
+    layout: LayoutJSON
+  ): Promise<LayoutMetadata> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${layoutsSaveLocation}`, {
+        headers: {
+          "Content-Type": "application/json",
+        },
+        method: "POST",
```
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:22]: I think I'm a bit of a strongly typed fiend, too used to Java, because I want to see this (and the others) as like `HttpRequest.POST`. I suppose this isn't a thing we can get out of the box with Typescript?

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  loadMetadata(): Promise<LayoutMetadata[]> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${metadataSaveLocation}`, {})
+        .then(async (response) => {
+          if (!response.ok) {
+            reject(new Error(response.statusText));
+          }
+          response.json().then((metadata: LayoutMetadata[]) => {
+            if (!metadata) {
+              reject(new Error("invalid metadata"));
```
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:116]:
```suggestion
reject(new Error("Response did not contain valid metadata"));
```

File: vuu-ui/packages/vuu-shell/src/layout-management/useLayoutManager.tsx [hunk 4]
```diff
@@ -34,25 +36,28 @@ export const LayoutManagementProvider = (props: LayoutManagementProviderProps) =
     persistenceManager.loadApplicationLayout().then(layout => {
       setApplicationLayout(layout);
     })
+      .catch((error: Error) => {
+        //TODO: Show error toaster
```
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-shell/src/layout-management/useLayoutManager.tsx:40]: :smile:

File: vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx [hunk 5]
```diff
@@ -116,7 +116,7 @@ const ShellWithNewTheme = () => {
   const { saveLayout } = useLayoutManager();
 
   const handleSave = useCallback(
-    (layoutMetadata: Omit<LayoutMetadata, "id">) => {
+    (layoutMetadata: Omit<LayoutMetadata, "id" | "created">) => {
```
Comment (cfisher-scottlogic) [inline vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx:119]: The fact we're having to do this omission of `id` and `created` anywhere we're sending a POST/PUT request makes me wonder whether we should have client-side DTOs as well. Not really a priority though, so unless it's a quick refactor it's not worth doing.

**Review by cfisher-scottlogic — APPROVED** [review]
LGTM! Only a couple minor points.

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+const metadataToAdd: Omit<LayoutMetadata, "id" | "created"> = {
+  name: "layout 1",
+  group: "group 1",
+  screenshot: "",
+  user: "username",
+};
+
+const layout: LayoutJSON = {
+  type: "",
+};
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts:33]: For the screenshots and type, I would put some dummy text in rather than empty strings. It might just safeguard against potential edge cases.

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+      expectPromiseRejectsWithError(
+        () => persistence.deleteLayout(uniqueId),
+        fetchError.message
+      );
+    });
+  });
+
+  describe("loadMetadata", () => {
+    it("resolves with array of metadata when response is ok", () => {
+      const persistence = new RemoteLayoutPersistenceManager();
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts:187]: Do we need to instantiate a separate persistence manager for each of the `loadMetadata`/`loadLayout` tests, or can we keep using the one in the outer scope?

File: .semgrepignore [hunk 8]
```diff
@@ -4,6 +4,7 @@ vuu/src/main/resources/www/ws-example.html
 vuu/src/main/scala/org/finos/vuu/provider/simulation/SimulatedBigInstrumentsProvider.scala
 vuu-ui/packages/vuu-data/src/array-data-source/group-utils.ts
 vuu-ui/packages/vuu-datagrid-extras/src/column-expression-input/column-language-parser/walkExpressionTree.ts
+vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply .semgrepignore:7]: [the static code analysis was failing in the pipeline](https://github.com/ScottLogic/finos-vuu/pull/74/checks?check_run_id=17732098110) beacuse we are using http instead of https. semgrp ignoring this file is not an ideal solution but will make the static code analysis pass for now. [this ticket](https://github.com/ScottLogic/finos-vuu/issues/79) will look into implementing SSL so that we don't need to do this. I have added a note to the ticket to remove this line from this file once we have SSL.

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  ): Promise<LayoutMetadata> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${layoutsSaveLocation}`, {
+        headers: {
+          "Content-Type": "application/json",
+        },
+        method: "POST",
+        body: JSON.stringify({
+          metadata: { ...metadata },
+          definition: JSON.stringify(layout),
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:25]: we don't need this stringify here as we are stringifying the whole body, will remove

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+  updateLayout(
+    id: string,
+    metadata: Omit<LayoutMetadata, "id">,
+    newLayoutJson: LayoutJSON
+  ): Promise<void> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${layoutsSaveLocation}/${id}`, {
+        method: "PUT",
+        body: JSON.stringify({
+          metadata,
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:54]: we dont need it in POST. will change. I was doing `metadata:{...metadata, screenshot: ""}` for testing when the screenshot string was giving us the issue with it being too long and I forgot to revert to just `metadata`

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  saveApplicationLayout(layout: LayoutJSON): Promise<void> {
+    // TODO POST api/layouts/application #71
+    console.log(layout);
+    return new Promise((resolve) => resolve());
+  }
+
+  loadApplicationLayout(): Promise<LayoutJSON> {
+    // TODO GET api/layouts/application #71
+    return new Promise((resolve) => resolve(defaultLayout));
+  }
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:136]: yes, it's #71

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts (hunk 6 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts:33]: good point. I will make the change

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts (hunk 7 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts:187]: we don't need a new instance here. Not sure why that's there. I will change

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts (hunk 1 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts]: applied. Thanks Cara :)

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+          definition: JSON.stringify(layout),
+        }),
+      })
+        .then((response) => {
+          if (!response.ok) {
+            reject(new Error(response.statusText));
+          }
+          response.json().then(({ metadata }: { metadata: LayoutMetadata }) => {
+            if (!metadata) {
+              reject(new Error("invalid metadata"));
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts]: applied. Thanks Cara :)

File: vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx (hunk 5 above)
Comment (vferraro-scottlogic) [reply vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx:119]: good point. I ended up using `Omit<LayoutMetadata, "id" | "created">` in a bunch of places and we might as well have it as its own type

Comment (vferraro-scottlogic) [reply vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx]: added a LayoutMetadataDto type for this

File: vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts [hunk 13]
```diff
@@ -10,7 +10,7 @@ export interface LayoutPersistenceManager {
    *
    * @returns Unique identifier assigned to the saved layout
    */
-  createLayout: (metadata: Omit<LayoutMetadata, "id">, layout: LayoutJSON) => Promise<string>;
+  createLayout: (metadata: Omit<LayoutMetadata, "id" | "created">, layout: LayoutJSON) => Promise<LayoutMetadata>;
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts]: yes I think it makes sense to have the save type here. I ll change it to use LayoutMetadataDto which omits "date" and "id"

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts (hunk 7 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts]: removed it! Thanks Peter :)

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    metadata: Omit<LayoutMetadata, "id" | "created">,
+    layout: LayoutJSON
+  ): Promise<LayoutMetadata> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${layoutsSaveLocation}`, {
+        headers: {
+          "Content-Type": "application/json",
+        },
+        method: "POST",
+        body: JSON.stringify({
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:23]: if we don't stringify the compiler complains with the following message 
```
Type '{ metadata: LayoutMetadataDto; definition: LayoutJSON; }' is not assignable to type 'BodyInit | null | undefined'.
      Object literal may only specify known properties, and 'metadata' does not exist in type 'ReadableStream<any> | Blob | ArrayBufferView | ArrayBuffer | FormData | URLSearchParams'.ts(2769)
```

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts (hunk 6 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts]: done!

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        })
+        .catch((error: Error) => {
+          reject(error);
+        })
+    );
+  }
+
+  loadLayout(id: string): Promise<LayoutJSON> {
+    return new Promise((resolve, reject) => {
+      fetch(`${baseURL}/${layoutsSaveLocation}/${id}`, {})
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:89]: that's correct "GET" is the default here. I don't mind making it explicit though if we think that makes the code more readable

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        .catch((error: Error) => {
+          reject(error);
+        });
+    });
+  }
+
+  loadMetadata(): Promise<LayoutMetadata[]> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${metadataSaveLocation}`, {})
+        .then(async (response) => {
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:110]: not needed. will remove

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts]: done

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts (hunk 2 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:22]: unfortunately the type definition that comes with 'node-fetch' package has type string for the request "method". It would have been nice if it was "GET" | "POST" | "PUT" and so on

Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:22]: we could implement our own fetch() and enforce more strict type for method if we wanted

Comment (cfisher-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:22]: I'm not concerned enough by it to force a refactor, seems like unnecessary additional code for a very small gain. Feel free to implement it if you want to.

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 17]
```diff
@@ -0,0 +1,137 @@
+import { LayoutMetadata } from "@finos/vuu-shell";
+import { LayoutPersistenceManager } from "./LayoutPersistenceManager";
+import { LayoutJSON } from "../layout-reducer";
+import { defaultLayout } from "./data";
+
+const baseURL = "http://127.0.0.1:8081/api";
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:6]: Something like having a BASE_URL environment variable and defaulting to "http://127.0.0.1:8081/api" if not defined?

File: vuu-ui/packages/vuu-layout/test/layout-persistence/LocalLayoutPersistenceManager.test.ts [hunk 18]
```diff
@@ -59,21 +70,25 @@ afterEach(() => {
...
-    const persistedMetadata = getLocalEntity<LayoutMetadata[]>(metadataSaveLocation);
+    const persistedMetadata =
+      getLocalEntity<LayoutMetadata[]>(metadataSaveLocation);
     const persistedLayout = getLocalEntity<Layout[]>(layoutsSaveLocation);
 
     const expectedMetadata: LayoutMetadata = {
       ...metadataToAdd,
-      id: returnedId
+      id,
+      created,
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/LocalLayoutPersistenceManager.test.ts:86]: done

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts (hunk 2 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:22]: discussed offline. no change needed

File: vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts]: I don't see any changes to `updateLayout` - do we want to use `LayoutMetadataDto` there too?

File: vuu-ui/packages/vuu-layout/test/layout-persistence/LocalLayoutPersistenceManager.test.ts [hunk 19]
```diff
@@ -59,23 +70,28 @@ afterEach(() => {
...
     };
 
     const expectedLayout: Layout = {
       json: layoutToAdd,
-      id: returnedId
+      id,
     };
 
+    expect(created).toEqual(formatDate(new Date(), "dd.mm.yyyy"));
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-layout/test/layout-persistence/LocalLayoutPersistenceManager.test.ts:94]:
```suggestion
    expect(created).toEqual(newDate);
```

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts (hunk 6 above)
Comment (pling-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts]: I think you've missed one for screenshot on line 18.

File: vuu-ui/packages/vuu-layout/src/layout-persistence/LocalLayoutPersistenceManager.ts [hunk 20]
```diff
@@ -1,48 +1,68 @@
...
     return new Promise((resolve, reject) => {
       this.validateIds(id)
         .then(() => Promise.all([this.loadLayouts(), this.loadMetadata()]))
         .then(([existingLayouts, existingMetadata]) => {
-          const layouts = existingLayouts.filter(layout => layout.id !== id);
-          const metadata = existingMetadata.filter(metadata => metadata.id !== id);
-          this.appendAndPersist(id, newMetadata, newLayout, layouts, metadata);
+          this.saveLayoutsWithMetadata(
+            existingLayouts.map((layout) =>
+              layout.id === id ? { ...layout, json: newLayout } : layout
```
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-layout/src/layout-persistence/LocalLayoutPersistenceManager.ts:57]: I'd find it easier to read with these arguments extracted into variables, but I like the overall change to mapping rather than filtering/appending.

File: vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts]: discussed offline: refactor so that we can use LayoutMetadataDto

File: vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts (hunk 6 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/RemoteLayoutPersistenceManager.test.ts]: done

File: vuu-ui/packages/vuu-layout/test/layout-persistence/LocalLayoutPersistenceManager.test.ts (hunk 19 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/test/layout-persistence/LocalLayoutPersistenceManager.test.ts]: done

File: vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/LayoutPersistenceManager.ts]: done!

File: vuu-ui/packages/vuu-layout/src/layout-persistence/LocalLayoutPersistenceManager.ts (hunk 20 above)
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/LocalLayoutPersistenceManager.ts]: change applied :)

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+        .then((response) => {
+          if (!response.ok) {
+            reject(new Error(response.statusText));
+          }
+          response.json().then((layout) => {
+            if (!layout) {
+              reject(new Error("invalid layout"));
+            }
+            resolve(layout);
+          });
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts]: will make the change to extract "definiton" from response.json() and will add the type

**Review by pling-scottlogic — APPROVED** [review]

File: vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+  ): Promise<LayoutMetadata> {
+    return new Promise((resolve, reject) =>
+      fetch(`${baseURL}/${layoutsSaveLocation}`, {
+        headers: {
+          "Content-Type": "application/json",
+        },
+        method: "POST",
+        body: JSON.stringify({
+          metadata,
+          definition: JSON.stringify(layout),
```
Comment (vferraro-scottlogic) [inline vuu-ui/packages/vuu-layout/src/layout-persistence/RemoteLayoutPersistenceManager.ts:30]: we do need this JSON.stringify() until we change the definition type #81
